dune/stuff/grid/periodicview.hh

```diff
@@ -0,0 +1,427 @@
+  EntityPointer outside_;
+  std::unique_ptr< DSC::ConstStorageProvider< RealGridViewType > > real_grid_view_;
+}; // ... class PeriodicIntersection ...
+
+/** \brief IntersectionIterator for PeriodicGridView
+ *
+ * PeriodicIntersectionIterator is derived from the IntersectionIterator of the underlying GridView and behaves exactly
+ * like the underlying IntersectionIterator except that it returns a PeriodicIntersection in its operator* and
+ * operator-> methods.
+ *
+ * \see PeriodicGridView
+ */
+template< class RealGridViewImp >
+class PeriodicIntersectionIterator
+    : public RealGridViewImp::IntersectionIterator
+{
+  typedef RealGridViewImp                                         RealGridViewType;
+  typedef typename RealGridViewType::IntersectionIterator         BaseType;
+
+public:
+  typedef typename BaseType::Intersection                         RealIntersectionType;
+  typedef int                                                     IntersectionIndexType;
+  typedef typename RealIntersectionType::EntityPointer            EntityPointerType;
+  typedef PeriodicIntersection< RealGridViewType >                Intersection;
+  typedef typename RealGridViewType::template Codim< 0 >::Entity  EntityType;
+  typedef std::pair< bool, EntityPointerType >                    PeriodicPairType;
+  static const size_t dimDomain = RealGridViewType::dimension;
+
+  PeriodicIntersectionIterator(BaseType real_intersection_iterator,
+                               const RealGridViewType& real_grid_view,
+                               const EntityType& entity,
+                               const std::map< IntersectionIndexType, PeriodicPairType >& intersection_map)
+    : BaseType(real_intersection_iterator)
+    , real_grid_view_(real_grid_view)
+    , entity_(entity)
+    , has_boundary_intersections_(entity_.hasBoundaryIntersections())
+    , intersection_map_(intersection_map)
+    , nonperiodic_pair_(std::make_pair(bool(false), EntityPointerType(entity_)))
+    , current_intersection_(create_current_intersection_safely())
+  {}
+
+  // methods that differ from BaseType
+  const Intersection& operator*() const
+  {
+    current_intersection_ = create_current_intersection();
+    return *current_intersection_;
+  }
+
+  const Intersection* operator->() const
+  {
+    current_intersection_ = create_current_intersection();
+    return &(*current_intersection_);
+  }
+
+private:
+  std::unique_ptr< Intersection > create_current_intersection() const
+  {
+    return DSC::make_unique< Intersection >(BaseType::operator*(),
+                                            real_grid_view_,
+                                            has_boundary_intersections_
+                                            ? intersection_map_.at((BaseType::operator*()).indexInInside())
+                                            : (const PeriodicPairType&)nonperiodic_pair_);
+  }
+
+  std::unique_ptr< Intersection > create_current_intersection_safely() const
```

Comment:
It is used in the constructor. We cannot use operator* if the iterator is equal to iend, but we need to initialize current_intersection_ nevertheless, so this method checks whether the iterator is iend and creates the intersection from ibegin in that case. For the methods operator* and operator->, this check is not needed, so I used the method create_current_intersection() there and create_current_intersection_safely() in the constructor.